### PR TITLE
feat(fastly): create fastly gateway to rotate between runtimes

### LINKED ISSUE #22: Include Backend URL in Gateway Response
let's say you have https://unviersal.runtime.com/helix-services/git-resolve-ref... and it is routed to AZURE and fails.
so it's probably important to include the backend request url in the response....

_Originally posted by @tripodsan in https://github.com/adobe/helix-deploy/issues/20#issuecomment-745001907_

### PR DESCRIPTION
**Pull request by trieloff**
depends on #19 for the `testURL` property

Fixes #22

### CODE REVIEW

**Comment by tripodsan** [comment]
why is this in helix-deploy ? the individual action deploys don't need to update the gateway. right?

one problem that might cause it to be very difficult to trace, is how to debug/trace which environment is used for a request.

let's say you have https://unviersal.runtime.com/helix-services/git-resolve-ref... and it is routed to AZURE and fails.
so it's probably important to include the backend request url in the response....

File: src/deploy/BaseDeployer.js [hunk 1]
```diff
@@ -27,6 +27,23 @@ class BaseDeployer {
   get relZip() {
     return path.relative(process.cwd(), this._builder.zipFile);
   }
+
+  get host() {
```
Comment (tripodsan) [inline src/deploy/BaseDeployer.js:31]: I don't think this is a good idea. 
the problem is, that the host is only known, once the action is deployed. but the host is dependent on the api gateway selected. which remains (should remain) stable. so better deduce the host from the gatway key or read it from AWS.

Comment (trieloff) [reply src/deploy/BaseDeployer.js:31]: I will update the AWSDeployer accordingly

Comment (trieloff) [reply src/deploy/BaseDeployer.js:31]: Fixed in 22cedb7

File: src/gateway/FastlyGateway.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+  selectBackendVCL() {
+    const vcl = `
+      declare local var.i INTEGER;
+      set var.i = randomint(0, ${this._deployers.length - 1});
+
+      if (1 == 0) {}`;
+
+    const middle = this._deployers.map((deployer, i) => `if(var.i <= ${i} && backend.F_${deployer.constructor.name.replace('Deployer', '')}.healthy) {
+      set req.backend = F_${deployer.constructor.name.replace('Deployer', '')};
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:62]:
```suggestion
      set req.backend = F_${deployer.name};
```

depends on #23

File: src/gateway/FastlyGateway.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }`);
+
+    const fallback = `{
+      set req.backend = F_${this._deployers[0].constructor.name.replace('Deployer', '')};
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:66]:
```suggestion
      set req.backend = F_${this._deployers[0].name};
```

#23

File: src/gateway/FastlyGateway.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      ${this._deployers[0].customVCL}
+    }`;
+
+    return [vcl, ...middle, fallback].join(' else ');
+  }
+
+  setURLVCL() {
+    return this._deployers.map((deployer) => `
+      if (req.backend == F_${deployer.constructor.name.replace('Deployer', '')}) {
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:75]:
```suggestion
      if (req.backend == F_${deployer.name}) {
```

File: src/gateway/FastlyGateway.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    await this._fastly.transact(async (newversion) => {
+      // set up health checks
+      await Promise.all(this._deployers
+        .map((deployer) => ({
+          check_interval: 60000,
+          expected_response: 200,
+          host: deployer.host,
+          http_version: '1.1',
+          initial: 1,
+          name: deployer.constructor.name.replace('Deployer', 'Check'),
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:93]:
```suggestion
          name: deployer.name +  'Check',
```

File: src/gateway/FastlyGateway.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+      // set up backends
+      await Promise.all(this._deployers
+        .map((deployer) => ({
+          hostname: deployer.host,
+          ssl_cert_hostname: deployer.host,
+          ssl_sni_hostname: deployer.host,
+          address: deployer.host,
+          name: deployer.constructor.name.replace('Deployer', ''),
+          healthcheck: deployer.constructor.name.replace('Deployer', 'Check'),
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:110]:
```suggestion
          name: deployer.name,
          healthcheck: deployer.name + 'Check',
```

**Comment by trieloff** [comment]
Added env vars to CircleCI, http://deploy-test.anywhere.run.global.prod.fastly.net/simple-name@1.45 can be tested

**Comment by trieloff** [comment]
@tripodsan https://2k3mydk3bl.execute-api.us-east-1.amazonaws.com/simple-package/simple-name_1.45 works, but https://2k3mydk3bl.execute-api.us-east-1.amazonaws.com/simple-package/simple-name_1.45/foo errors out – not sure why, but feel free to take over.

**Comment by trieloff** [comment]
@tripodsan I've created a service at https://manage.fastly.com/configure/services/2oGtQaKJnfHAv5jyCpxhEd/versions/1/domains that we can use for helix-status when the AWS API Gateway account is ready.

**Comment by trieloff** [comment]
Rotate between AWS + OW: https://deploy-test.anywhere.run/simple-name@1.45/bar3

File: src/action_builder.js [hunk 7]
```diff
@@ -936,10 +955,19 @@ module.exports = class ActionBuilder {
 
     await this.updateLinks();
 
+    if (this._gateways.fastly && this._gateways.fastly.ready()) {
+      Object.values(this._deployers).forEach((d) => {
+        this._gateways.fastly.withDeployer(d);
+      });
+
+      this._gateways.fastly.init();
+      await this._gateways.fastly.deploy();
```
Comment (tripodsan) [inline src/action_builder.js:964]: @trieloff shouldn't we put this behind another _command_ ? like `--update-gateway` 
I find the precense of the gateway a weak trigger, especially if the env vars are always there in a ci.

Comment (trieloff) [reply src/action_builder.js:964]: Good idea

File: src/gateway/FastlyGateway.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        type: 'miss',
+        content: this.setURLVCL(),
+      });
+
+      await this._fastly.writeSnippet(newversion, 'logurl', {
+        name: 'logurl',
+        priority: 10,
+        dynamic: 0,
+        type: 'fetch',
+        content: `set beresp.http.X-Backend-URL = bereq.url;
```
Comment (trieloff) [inline src/gateway/FastlyGateway.js:183]: @tripodsan this is the fix for #22

**Pull request merged**